src/umzug.ts

```diff
@@ -1,416 +1,353 @@
+			force?: boolean;
+
+			/** Should not be specified with `migrations` */
+			to?: never;
+	  };
+
+export type MigrateDownOptions =
+	| {
+			/** If specified, migrations down to and including this name will be revert. Otherwise, only the last executed will be reverted */
+			to?: string | 0;
+
+			/** Should not be specified with `to` */
+			migrations?: never;
+
+			/** Should not be specified with `to` */
+			force?: never;
+	  }
+	| {
+			/** If specified, only these migrations will be reverted. An error will be thrown if any of the names are not found in the list of executed migrations */
+			migrations: string[];
+
+			/** Allow reverting migrations which have not been run yet. Use with caution. */
+			force?: boolean;
+
+			/** Should not be specified with `migrations` */
+			to?: never;
+	  };
+
+export class Umzug<Ctx> extends EventEmitter {
+	private readonly storage: UmzugStorage;
+	private readonly migrations: () => Promise<readonly RunnableMigration[]>;
 
-	constructor(options?: UmzugConstructorOptions) {
+	/**
+	 * Compile-time only property for type inference. After creating an Umzug instance, it can be used as type alias for
+	 * a user-defined migration. The function receives a migration name, path and the context for an umzug instance
+	 * @example
+	 * // migrator.ts
+	 * import { Umzug } from 'umzug'
+	 *
+	 * const umzug = new Umzug({...})
+	 * export type Migration = typeof umzug._types.migration;
+	 *
+	 * umzug.up();
+	 *
+	 * @example
+	 * // migration-1.ts
+	 * import type { Migration } from '../migrator'
+	 *
+	 * // name and context will now be strongly-typed
+	 * export const up: MigrationFn = ({name, context}) => context.query(...)
+	 * export const down: MigrationFn = ({name, context}) => context.query(...)
+	 */
+	readonly _types: {
+		migration: (params: { name: string; path?: string; context: Ctx }) => Promise<unknown>;
+	} = {} as any;
+
+	/** creates a new Umzug instance */
+	constructor(private readonly options: UmzugOptions<Ctx>) {
 		super();
-		options = options ?? {};
 
-		if (options.logging && typeof options.logging !== 'function') {
-			throw new Error('The logging-option should be either a function or false');
-		}
+		this.storage = verifyUmzugStorage(options.storage || new JSONStorage());
+		this.migrations = this.getMigrationsResolver();
+	}
 
-		let migrations;
-		if (Array.isArray(options.migrations)) {
-			migrations = options.migrations;
-		} else {
-			migrations = {
-				params: [],
-				path: path.resolve(process.cwd(), 'migrations'),
-				pattern: /^\d+[\w-]+\.js$/,
-				traverseDirectories: false,
-				wrap: (fn: () => Promise<any>) => fn,
-				...options.migrations,
-			};
+	private logging(message: string) {
+		this.options.logger?.info(message);
+	}
+
+	static defaultResolver: Resolver<unknown> = ({ path: filepath, name, context }) => {
+		const ext = path.extname(filepath);
+		const canRequire = ext === '.js' || ext in require.extensions;
```

Comment:
Yeah, this is just a UX optimisation. `ts-node` adds `'.tsc` to `require.extensions`, so this is a cheaper/cleaner way to check that we'll be able to `require` the file than try/catch. I doubt it will be removed any time soon since ts-node is quite widely used. If/when it's removed it'd affect ts-node first, so we'd probably want to look at however ts-node updates its `ts-node/register` module.